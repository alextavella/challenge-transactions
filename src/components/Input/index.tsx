import MaterialTextField, {
  BaseTextFieldProps,
} from '@material-ui/core/TextField';
import React from 'react';

export interface InputProps extends BaseTextFieldProps {
  name: string;
  label: string;
  type?: string;
  defaultValue?: string;
}

const Input: React.ForwardRefRenderFunction<any, InputProps> = (
  { name, label, type = 'text', defaultValue = '', ...rest },
  ref,
) => {
  return (
    <MaterialTextField
      {...rest}
      inputRef={ref}
      name={name}
      label={label}
      type={type}
      defaultValue={defaultValue}
      variant="outlined"
      InputProps={{
        inputProps: { 'data-testid': name },
      }}
    />
  );
};

export default React.forwardRef(Input);
