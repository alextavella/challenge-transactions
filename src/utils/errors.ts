import { ValidationError } from 'yup';

interface Error {
  [key: string]: string;
}

export function getValidationErrors(errors: ValidationError): any {
  return errors.inner.reduce((acc: Error, error: ValidationError) => {
    return { ...acc, [error.path]: error.message };
  }, {});
}
