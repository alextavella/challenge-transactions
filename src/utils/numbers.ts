export const formatCurrency = (value: number): string => {
  const formatted = new Intl.NumberFormat([], {
    style: 'currency',
    currency: 'BRL',
    minimumFractionDigits: 2,
  }).format(value);

  return formatted
    .replace('.', '#')
    .replace(/,/g, '.')
    .replace('#', ',')
    .replace('R$', 'R$ ');
};

export const unformatCurrency = (value: string): number => {
  return Number(
    value
      .replace('R$ ', '')
      .replace(',', '#')
      .replace(/\./g, '')
      .replace('#', '.'),
  );
};
