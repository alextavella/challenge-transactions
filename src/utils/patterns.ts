import { ValidationRules } from 'react-hook-form';

export interface MaskRule {
  rules: ValidationRules;
  transform: (text: string) => string;
}

export const CPF: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{3}).([0-9]{3}).([0-9]{3})-([0-9]{2})$/),
      message: 'CPF inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,3})(\d{0,3})(\d{0,3})(\d{0,2})/)!;

    const part1 = result[2] ? `${result[1]}.${result[2]}` : result[1];
    const part2 = result[3] ? `.${result[3]}` : '';
    const part3 = result[4] ? `-${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_NUMBER: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{4})\s([0-9]{4})\s([0-9]{4})\s([0-9]{4})$/),
      message: 'Número do cartão inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,4})(\d{0,4})(\d{0,4})(\d{0,4})/)!;

    const part1 = result[2] ? `${result[1]} ${result[2]}` : result[1];
    const part2 = result[3] ? ` ${result[3]}` : '';
    const part3 = result[4] ? ` ${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_EXPIRATION_DATE: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^(0[1-9]|1[012])\/?(20[2-9][0-9])$/),
      message: 'Data de expiração inválida',
    },
  },
  transform: (text: string = ''): string => {
    const result = text.replace(/\D/g, '').match(/(\d{0,2})(\d{0,4})/)!;

    return result[2] ? `${result[1]}/${result[2]}` : result[1];
  },
};

export const CREDIT_CARD_CVV: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/[0-9]{3}/),
      message: 'CVV inválido',
    },
  },
  transform: (text: string = ''): string => {
    return text.substr(0, 3);
  },
};

export const CURRENCY: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/([0-9]\.?)?[0-9]+,[0-9]{2}/),
      message: 'Valor inválido',
    },
  },
  transform: (text: string = ''): string => {
    let number = text.replace(/[^0-9]/g, '');

    if (number.length === 0) number = '0.00';
    else if (number.length === 1) number = '0.0' + number;
    else if (number.length === 2) number = '0.' + number;
    else
      number =
        number.substring(0, number.length - 2) +
        '.' +
        number.substring(number.length - 2, number.length);

    number = Number(number).toFixed(2);
    number = number.replace(/\./g, ',');

    const x = number.split(',');
    const x1 = x[0];
    const x2 = x.length > 1 ? ',' + x[1] : '';

    const formatAmountNoDecimals = (number: string) => {
      var rgx = /(\d+)(\d{3})/;
      while (rgx.test(number)) {
        number = number.replace(rgx, `$1.$2`);
      }
      return number;
    };

    return `R$ ` + formatAmountNoDecimals(x1) + x2;
  },
};
