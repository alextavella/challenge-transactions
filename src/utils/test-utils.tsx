import { render } from '@testing-library/react';
import React from 'react';
import MainProvider from '../providers';

const AllTheProviders: React.FC = ({ children }) => {
  return <MainProvider>{children}</MainProvider>;
};

const customRender = (ui: React.ReactElement, options?: any) =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
