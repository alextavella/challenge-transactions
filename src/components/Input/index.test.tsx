import React from 'react';
import Input from '.';
import { cleanup, fireEvent, render, screen } from '../../utils/test-utils';

describe('<Input />', () => {
  afterEach(() => {
    cleanup();
  });

  it('should be able to render correctly', () => {
    // Arrange
    const { baseElement } = render(<Input name="nome" label="Nome" />);

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to render correctly when disabled', () => {
    // Arrange
    const { baseElement } = render(<Input name="nome" label="Nome" disabled />);

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to render correctly with default value', () => {
    // Arrange
    render(<Input name="name" label="name" defaultValue="Alex Tavella" />);

    // Assert
    expect(screen.getByDisplayValue('Alex Tavella')).toBeTruthy();
  });

  it('should be able to change text', () => {
    // Arrange
    render(<Input name="name" label="name" />);

    const input = screen.getByTestId('name');

    // Act
    fireEvent.change(input, { target: { value: 'Alex Tavella' } });

    // Assert
    expect(screen.getByDisplayValue('Alex Tavella')).toBeTruthy();
  });
});
