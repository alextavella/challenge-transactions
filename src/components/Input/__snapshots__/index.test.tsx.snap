// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should be able to render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Nome
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          data-testid="nome"
          name="nome"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              Nome
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</body>
`;

exports[`<Input /> should be able to render correctly when disabled 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled"
        data-shrink="false"
      >
        Nome
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
          data-testid="nome"
          disabled=""
          name="nome"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-7"
          >
            <span>
              Nome
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</body>
`;
